### PR TITLE
Add single request timeout to the contract

### PR DESCRIPTION
**Pull request by slinkydeveloper**
Signed-off-by: Francesco Guardiani <francescoguard@gmail.com>

This adds the support for the experimental feature `delivery-timeout`: https://github.com/knative/eventing/issues/5148

Also part of what was discussed in #757

## Proposed Changes

- :gift: Add `EgressConfig.timeout` to configure the single request timeout

### CODE REVIEW

File: data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/http/HttpConsumerVerticleFactory.java [hunk 1]
```diff
@@ -228,23 +228,34 @@ private ConsumerRecordSender createConsumerRecordSender(
...
+      options.setTimeout(
+        egressConfig.getTimeout() > 0 ?
+          egressConfig.getTimeout() :
+          CircuitBreakerOptions.DEFAULT_TIMEOUT
+      );
+
+      // Retry options
+      if (egressConfig.getRetry() > 0) {
+        return new CircuitBreakerOptions()
+          // TODO reset timeout should be configurable or, at least, set by the control plane
```
Comment (slinkydeveloper) [inline data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/http/HttpConsumerVerticleFactory.java:247]: @pierDipi does this comment still applies?

File: data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/http/HttpConsumerVerticleFactory.java [hunk 2]
```diff
@@ -228,23 +228,34 @@ private ConsumerRecordSender createConsumerRecordSender(
...
+      // Single request timeout
+      options.setTimeout(
+        egressConfig.getTimeout() > 0 ?
+          egressConfig.getTimeout() :
+          CircuitBreakerOptions.DEFAULT_TIMEOUT
+      );
+
+      // Retry options
+      if (egressConfig.getRetry() > 0) {
+        return new CircuitBreakerOptions()
```
Comment (matzew) [inline data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/http/HttpConsumerVerticleFactory.java:246]: why do we return here a new object?

Comment (slinkydeveloper) [reply data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/http/HttpConsumerVerticleFactory.java:246]: Wops my bad!

**Comment by matzew** [comment]
Process is:
1. edit the `proto/contract.proto`
2. call the `generate-proto.sh` script (which updates the two generated files)
3. manual update the factory here, to make use of the timeout

 Right?

**Comment by slinkydeveloper** [comment]
That's exactly what I did for this PR

**Comment by matzew** [comment]
LGTM

File: data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/http/HttpConsumerVerticleFactory.java (hunk 1 above)
Comment (pierDipi) [reply data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/http/HttpConsumerVerticleFactory.java:247]: Yes, let's keep it

File: proto/contract.proto [hunk 3]
```diff
@@ -46,6 +46,9 @@ message EgressConfig {
 
   // backoffDelay is the delay before retrying in milliseconds.
   uint64 backoffDelay = 4;
+
+  // timeout is the single request timeout (not the overall retry timeout)
+  uint64 timeout = 5;
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply proto/contract.proto:51]: I know, not in this PR though :)

[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply proto/contract.proto:51]: We need to merge delivery-timeout first https://github.com/knative/eventing/issues/5148

**Pull request merged**